File: leetcode/src/array/P_17_21.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        int count = 0, sum = 0, i = 0, j = height.size() - 1, k = 1;
        for(auto t = height.begin(); t < height.end(); t++) {
            sum = sum + *t;
        }
        while (i < j) {
            if (height[i] != k) {
                i++;
            }
            if (height[j] != k) {
                j--;
            }

            if (height[i] == height[j] && height[i] == k) {
                count = count + (j - i);
                i++;
                j--;
                k++;
            }
        }
        return sum - count;
    }
};

File: leetcode/src/array/P_17_21.java
package array;

public class P_17_21 {
    public int trap(int[] height) {
        int count = 0, sum = 0, i = 0, j = height.length - 1, k = 1;
        for (int value : height) {
            sum = sum + value;
        }
        while (i <= j) {
            while (i <= j && height[i] < k) {
                i++;
            }
            while (i <= j && height[j] < k) {
                j--;
            }
            count = count + j - i + 1;
            k++;
        }
        return count - sum;
    }
}
